src/trampoline.rs

```diff
@@ -0,0 +1,86 @@
+use neon_runtime::raw::Env;
+use neon_runtime::tsfn::{CallError, ThreadsafeFunction};
+
+use crate::context::TaskContext;
+use crate::result::NeonResult;
+
+pub(crate) type Callback = Box<dyn FnOnce(Env) + Send + 'static>;
+
+pub(crate) struct ThreadsafeTrampoline {
+    tsfn: ThreadsafeFunction<Callback>,
+    ref_count: u32,
+}
+
+impl ThreadsafeTrampoline {
+    /// Creates an unbounded queue for scheduling closures on the JavaScript
+    /// main thread
+    pub(crate) fn new(env: Env) -> Self {
+        let tsfn = unsafe { ThreadsafeFunction::new(env, Self::callback) };
+
+        Self {
+            tsfn: tsfn,
+            ref_count: 1,
+        }
+    }
+
+    /// Schedules a closure to execute on the JavaScript thread that created
+    /// this ThreadsafeTrampoline.
+    /// Returns an `Error` if the task could not be scheduled.
+    pub(crate) fn try_send<F>(&self, f: F) -> Result<(), CallError<Callback>>
+    where
+        F: FnOnce(TaskContext) -> NeonResult<()> + Send + 'static,
+    {
+        let callback = Box::new(move |env| {
+            let env = unsafe { std::mem::transmute(env) };
+
+            // Note: It is sufficient to use `TaskContext`'s `InheritedHandleScope` because
+            // N-API creates a `HandleScope` before calling the callback.
+            TaskContext::with_context(env, move |cx| {
+                let _ = f(cx);
+            });
+        });
+
+        self.tsfn.call(callback, None)
+    }
+
+    /// References a trampoline to prevent exiting the event loop until it has been dropped. (Default)
+    /// Safety: `Env` must be valid for the current thread
+    /// _Not idempotent_
+    pub(crate) fn reference(&mut self, env: Env) {
```

Comment:
Nitpick: Maybe we should make this more obviously distinct from the idempotent nature of `ref` by calling it `increment_reference_count` or something.